crypto/x509/v3_akid.c

```diff
@@ -156,53 +155,38 @@ static AUTHORITY_KEYID *v2i_AUTHORITY_KEYID(X509V3_EXT_METHOD *method,
         ERR_raise(ERR_LIB_X509V3, X509V3_R_NO_ISSUER_CERTIFICATE);
         goto err;
     }
-    same_issuer = ctx->subject_cert == ctx->issuer_cert;
-    ERR_set_mark();
-    if (ctx->issuer_pkey != NULL)
-        ss = X509_check_private_key(ctx->subject_cert, ctx->issuer_pkey);
-    else
-        ss = same_issuer;
-    ERR_pop_to_mark();
-
-    /* unless forced with "always", AKID is suppressed for self-signed certs */
-    if (keyid == 2 || (keyid == 1 && !ss)) {
+
+    /* unless forced with "always", AKID is optional */
+    if (keyid != 0) {
         /*
          * prefer any pre-existing subject key identifier of the issuer cert
-         * except issuer cert is same as subject cert and is not self-signed
+         * except issuer cert is same as subject cert and private key is given
          */
-        i = X509_get_ext_by_NID(issuer_cert, NID_subject_key_identifier, -1);
-        if (i >= 0 && (ext = X509_get_ext(issuer_cert, i)) != NULL
-            && !(same_issuer && !ss)) {
-            ikeyid = X509V3_EXT_d2i(ext);
-            if (ASN1_STRING_length(ikeyid) == 0) /* indicating "none" */ {
-                ASN1_OCTET_STRING_free(ikeyid);
-                ikeyid = NULL;
-            }
-        }
-        if (ikeyid == NULL && same_issuer && ctx->issuer_pkey != NULL) {
+        if (ctx->subject_cert == issuer_cert && ctx->issuer_pkey != NULL) {
```

Comment:
Your code did not match the (sensible) behavior you documented. 
Need to this by really preferring any existing SKID:
```suggestion
        i = X509_get_ext_by_NID(issuer_cert, NID_subject_key_identifier, -1);
        if (i >= 0 && (ext = X509_get_ext(issuer_cert, i)) != NULL) {    
            ikeyid = X509V3_EXT_d2i(ext);
        } else if (ctx->subject_cert == issuer_cert && ctx->issuer_pkey != NULL) {
```

Reply:
Why do you want to change the semantics of SKID addition?
And if this really needs to be done, better not do a lib level something different form app level.

Reply:
I want to simplify this, the private key shall only be set by the application, when
both the certificate is SELF-ISSUED and the public key is not matching the private key.
There shall be no cryptographic operation necessary in the config parser.
And the private key shall be only used as the keyid OVERRIDE, it is the decision of the
application if the OVERRIDE is necessary, it is rather an exception.
Current implementation enforce the self-signed rules also for self-issued proxy certificates
when the subject name does not match the issuer name, but just the key is re-used.
At this part the backport PRs #29092 for 3.6 and #29079 for 3.3 are slightly different
than this PR: they do not have this commit: 6715dde4ad64d667c21db2354f5b6045ce5c9b52
This is of course clearly not okay for a back-port to a stable version.
And yes, the high-level API is in total control of what the low-level API does.